ratis-server/src/main/java/org/apache/ratis/server/impl/ServerState.java

```diff
@@ -318,19 +318,24 @@ boolean recognizeCandidate(RaftPeerId candidateId, long candidateTerm) {
     return false;
   }
 
-  boolean isLogUpToDate(TermIndex candidateLastEntry) {
+  int compareLog(TermIndex candidateLastEntry) {
     TermIndex local = log.getLastEntryTermIndex();
     // need to take into account snapshot
     SnapshotInfo snapshot = server.getStateMachine().getLatestSnapshot();
-     if (local == null && snapshot == null) {
-      return true;
+    if (local == null && snapshot == null) {
+      // If the lastEntry of candidate is null, the proto will transfer an empty TermIndexProto,
+      // then term and index of candidateLastEntry will both be 0
+      if (candidateLastEntry.getTerm() == 0 && candidateLastEntry.getIndex() == 0) {
+        return 0;
+      }
+      return -1;
     } else if (candidateLastEntry == null) {
```

Comment:
The case candidateLastEntry == null is never happen so that  we should remove it.  (Or, if we want to support candidateLastEntry == null, we have to check null in line 328.)